rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -150,3 +157,75 @@ impl<Protocol> IntoResponse<Protocol> for MissingServerRequestId {
         internal_server_error()
     }
 }
+
+#[derive(Clone)]
+pub struct ServerRequestIdResponseProvider<S> {
+    inner: S,
+    header_key: HeaderName,
+}
+
+/// A layer that inserts the generated [`ServerRequestId`] to the response headers
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct ServerRequestIdResponseProviderLayer {
+    header_key: Box<str>,
+}
+
+impl ServerRequestIdResponseProviderLayer {
+    /// Add the request ID to the response header `header_key`
+    pub fn new(header_key: Box<str>) -> Self {
+        Self { header_key }
+    }
+}
+
+impl<S> Layer<S> for ServerRequestIdResponseProviderLayer {
+    type Service = ServerRequestIdResponseProvider<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        let header_key = HeaderName::from_bytes(self.header_key.as_bytes()).unwrap();
+        ServerRequestIdResponseProvider { inner, header_key }
+    }
+}
+
+impl<Body, S> Service<http::Request<Body>> for ServerRequestIdResponseProvider<S>
+where
+    S: Service<http::Request<Body>, Response = Response<crate::body::BoxBody>>,
+    S::Future: std::marker::Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = std::pin::Pin<Box<dyn Send + std::future::Future<Output = Result<Self::Response, Self::Error>>>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: http::Request<Body>) -> Self::Future {
+        let request_id = req.extensions().get::<ServerRequestId>();
+        if request_id.is_none() {
+            return Box::pin(self.inner.call(req));
+        }
```

Comment:
I think we should go in the opposite direction - panic with an error message that explains that they likely misconfigured their server and that the other layer must be registered before this one.
Silent failures are the worst.

Reply:
Fair. I would have preferred to check this at compile time, but I don't think it's currently possible here

Reply:
No, I don't think you can (in any reasonable way).

Reply:
How do you feel about only implementing `Service` when the inner service is `ServerRequestIdProvider<S>`? This would force users to apply `ServerRequestId` then `ServerResponseId` else they'd get a compile error.

Reply:
What does the compiler error look like?

Reply:
The change is on `ServerRequestIdProvider`. I've left the implementation [here](https://github.com/awslabs/smithy-rs/pull/2438/commits/147eef26d5f8972b810dc5994e57865a872e44b0). If I swap the layers incorrectly, the errors are:
```
error[E0277]: the trait bound `ServerRequestIdProvider<Route<_>>: Service<Request<Body>>` is not satisfied
  --> pokemon-service/src/bin/pokemon-service.rs:63:51
   |
63 |     let server = hyper::Server::bind(&bind).serve(app.into_make_service());
   |                                             ----- ^^^^^^^^^^^^^^^^^^^^^^^ the trait `Service<Request<Body>>` is not implemented for `ServerRequestIdProvider<Route<_>>`
   |                                             |
   |                                             required by a bound introduced by this call
   |
   = help: the trait `Service<Request<Body>>` is implemented for `ServerRequestIdProvider<ServerRequestIdResponseProvider<S>>`
   = note: required because of the requirements on the impl of `Service<Request<Body>>` for `ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>`
   = note: required because of the requirements on the impl of `hyper::service::http::HttpService<Body>` for `PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>`
   = note: required because of the requirements on the impl of `hyper::service::make::MakeServiceRef<AddrStream, Body>` for `IntoMakeService<PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>>`

error[E0277]: the trait bound `hyper::common::exec::Exec: hyper::common::exec::ConnStreamExec<RoutingFuture<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>, Body>, http_body::combinators::box_body::UnsyncBoxBody<hyper::body::Bytes, aws_smithy_http_server::error::Error>>` is not satisfied
  --> pokemon-service/src/bin/pokemon-service.rs:66:29
   |
66 |     if let Err(err) = server.await {
   |                             ^^^^^^ the trait `hyper::common::exec::ConnStreamExec<RoutingFuture<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>, Body>, http_body::combinators::box_body::UnsyncBoxBody<hyper::body::Bytes, aws_smithy_http_server::error::Error>>` is not implemented for `hyper::common::exec::Exec`
   |
   = help: the trait `hyper::common::exec::ConnStreamExec<F, B>` is implemented for `hyper::common::exec::Exec`
   = note: required because of the requirements on the impl of `Future` for `Server<AddrIncoming, IntoMakeService<PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>>>`
   = note: required because of the requirements on the impl of `IntoFuture` for `Server<AddrIncoming, IntoMakeService<PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>>>`
help: remove the `.await`
   |
66 -     if let Err(err) = server.await {
66 +     if let Err(err) = server {
   |

error[E0277]: the trait bound `hyper::common::exec::Exec: hyper::common::exec::NewSvcExec<AddrStream, aws_smithy_http_server::routing::into_make_service::MakeRouteServiceFuture<PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>>, PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>, hyper::common::exec::Exec, hyper::server::server::NoopWatcher>` is not satisfied
  --> pokemon-service/src/bin/pokemon-service.rs:66:29
   |
66 |     if let Err(err) = server.await {
   |                             ^^^^^^ the trait `hyper::common::exec::NewSvcExec<AddrStream, aws_smithy_http_server::routing::into_make_service::MakeRouteServiceFuture<PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>>, PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>, hyper::common::exec::Exec, hyper::server::server::NoopWatcher>` is not implemented for `hyper::common::exec::Exec`
   |
   = help: the trait `hyper::common::exec::NewSvcExec<I, N, S, E, W>` is implemented for `hyper::common::exec::Exec`
   = note: required because of the requirements on the impl of `Future` for `Server<AddrIncoming, IntoMakeService<PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>>>`
   = note: required because of the requirements on the impl of `IntoFuture` for `Server<AddrIncoming, IntoMakeService<PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>>>`
help: remove the `.await`
   |
66 -     if let Err(err) = server.await {
66 +     if let Err(err) = server {
   |

error[E0277]: the trait bound `ServerRequestIdProvider<Route<_>>: Service<Request<Body>>` is not satisfied
  --> pokemon-service/src/bin/pokemon-service.rs:63:45
   |
63 |     let server = hyper::Server::bind(&bind).serve(app.into_make_service());
   |                                             ^^^^^ the trait `Service<Request<Body>>` is not implemented for `ServerRequestIdProvider<Route<_>>`
   |
   = help: the trait `Service<Request<Body>>` is implemented for `ServerRequestIdProvider<ServerRequestIdResponseProvider<S>>`
   = note: required because of the requirements on the impl of `Service<Request<Body>>` for `ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>`
   = note: required because of the requirements on the impl of `hyper::service::http::HttpService<Body>` for `PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>`
   = note: required because of the requirements on the impl of `hyper::service::make::MakeServiceRef<AddrStream, Body>` for `IntoMakeService<PokemonService<AddExtension<ServerRequestIdResponseProvider<ServerRequestIdProvider<Route<_>>>, Arc<State>>>>`
``` 

These are the layers in the example:
```
        .layer(&ServerRequestIdProviderLayer::new()) /* Generate a server request ID */
        .layer(&ServerRequestIdResponseProviderLayer::new("X-Request-Id".into())) /* Add the server generated ID to the response */
        .layer(&AddExtensionLayer::new(Arc::new(State::default())));
```

Reply:
OK, quite disgusting. My preference is for a runtime error, but feel free to move forward with what you think is best.

Reply:
I don't know how much we care, but then you'd have to add `ServerRequestIdResponseProviderLayer`

Reply:
```rust
impl<Body, S> Service<http::Request<Body>> for ServerRequestIdResponseProvider<ServerRequestIdProvider<S>> {
```
is the wrong order also - we need to add a request ID via `ServerRequestIdProvider` before `ServerRequestIdResponseProvider`. A implementation for `ServerRequestIdProvider<ServerRequestIdResponseProvider<S>>` would force them to always add response IDs, which is not what we want.